### ISSUE TITLE
Ability to modify entries in KV

### ISSUE THREAD

**Issue by CodeSandwich** [slog-rs/slog#211]
# Context
I have application divided into sections. Every section name is in logger's `KV`, so log consumer can tell which section is source of entry. I want to pass logger from a section to a sub-section and set a new section name in the `KV` to reflect that.

# Problem
I can't get rid of the old section name in sub-section:
- `KV` and `OwnedKVList` have no API for modification
- `Logger` can't have its drain extracted
- `Logger::new` preserves whole old `KV`
- `Logger` with two `KV`s outputs all entries from both of them. There is no overriding or shadowing. This causes problems with JSON output formats, which generate invalid structures with duplicate keys (I'm looking at you, slog-json)

# Solutions
- Extend `OwnedKVList` API to have means of editing. Because of generic nature this will cause same change in `KV`, which in turn will make a huge breaking change in API.
- Add method to extract `Drain` from `Logger`. That actually looks easy, but is least powerful, allows only full clearing of `KV`.
- Make `Logger` iterate over al `KV`s on all layers of wrapping and remove duplicates with preference of values less shallowly in wrapped.

**Comment by dpc** [comment]
Another solution is to implement a `Drain` that removes duplicates. This one is good because everyone can implement it on their own, and since they know which keys can change, they can optimize it a lot over a general solution.

It seems to me that it should be possible to have a function that takes old `OwnedKVList` and assembles a new, modified version. That can be passed to `Logger::root` to create a new `Logger` with replaced values, but same `Drain`. Maybe we should even add `Logger::new_replace` or something like this. I like this one.

**Comment by CodeSandwich** [comment]
Is it possible with current API? AFAIK `Drain::log` receives `&OwnedKVList`, which can't do anything with its public API. This leads us back to my proposition 1, extending `OwnedKVList`'s API.

**Comment by dpc** [comment]
@CodeSandwich Yes it is. You can serialize it into arbitrary `Serializer` https://docs.rs/slog/2.4.1/slog/trait.KV.html#tymethod.serialize and in that `Serializer` you can keep track of, skip, replace,  and whatever you want with each key-value pair.

**Comment by CodeSandwich** [comment]
So if I understand correctly, the idea is to create `Drain` which on every log will create a fake serializer and pass it to `KV`. This serializer will intercept and store all keys and filter out duplicates based on their values. Finally `Drain` will extract stored mapping and pass it as `&OwnedKVList` down to its child.

The problem is that it's not possible to to create an `OwnedKVList` without macro and macro requires to know all keys during compilation.

This could be possible to use, but personally I would prefer to avoid it at any cost. The implementation looks moderately hard in single specific case. It also seems extremely hard to make configurable, probably possible only with some macro generating custom types. It also almost certainly involves allocation of collection on every log event.

**Comment by CodeSandwich** [comment]
I think that adding possibility to construct `OwnedKV` by hand could solve this issue. This is a simple newtype and a struct, so it would be 2 lines of code without risking breaking changes.

It would make it possible to create all kind of custom elements. I'm thinking about a custom `DedupLogger` which would store a `HashMap` of key-values along a copy of them in an `OwnedKV`. When sombody tries to wrap it in a new `DedupLogger`, they could agree on new set of key-values and construct new `HashMap` and `OwnedKV` to use by the parent. Because logging is always performed with macros, which are defacto duck-typed, `DedupLogger` could be used instead of `Logger` if only it had the same definition of `log` method. Not a very pretty solution, but definetely least effort on slog side.

**Comment by dpc** [comment]
We could add some constructors to `OwnedKVList`.

**Comment by CodeSandwich** [comment]
Why `OwnedKVList` and not `OwnedKV`? The latter is more generic and can generate `OwnedKVList`, but not the other way around.

**Comment by eranrund** [comment]
Making the `Drain` able to modify `KV` would also be useful to have a drain that includes the current thread name, which is a thing I've been trying to do unsuccessfully with the current API.

**Comment by dpc** [comment]
Just a thought - there's nothing preventing creation of a new `Logger` with a `KV` inside entirely rebuild. One could just iterate through all the KVs and clone them, but replace/insert new ones etc.

**Comment by dpc** [comment]
@eranrund What's the problem with current-thread? Is it async logger?

**Comment by eranrund** [comment]
@dpc It is `async`...

**Comment by dpc** [comment]
The missing piece of early evaluation. One approach would be to add a new field in the `Record` which would be "early-evaluated-owned-kvs". Then couple of functions to add them to `Logger. `slog-async` would evaluate them early and pass further just like borrowed values. I think it would be a backward compatible change, and quite easy to make.

Eventually - it could be just added to `slog-async` itself. Just register KVs that `slog-async` will add. It makes sense, I guess as `async` is the only time this feature is going to be needed.

**Comment by eranrund** [comment]
I like that suggestion. Is that something you expect to have time and interest to build in the near future?

**Comment by dpc** [comment]
@eranrund `slog` is now entirely in "you fill PRs for what you need" mode. :D

**Comment by eranrund** [comment]
Fair! I'll look into doing that sometime soon, I hope. I'm in the process of converting a fairly large internal codebase to using `slog`, so I want to get us committed to using it and then I could implement this. Thank you!

**Comment by dpc** [comment]
> I had the idea to create my own `Drain` that could wrap `slog_bunyan` and handle duplicates based on user wishes (pass an enum like `DiscardDuplicates`, `LeaveAsIs`, `ConvertToArray`).

That is probably the way to go about it.

slog key-value handling is optimized to make it as lightweight as possible, so the key values are (be default) static strings, and are chained together to avoid needing to allocate a container for them.

These few rare use-cases, while valid, are uncommon enough to push the extra work on people who need it.
